module/executiondatasync/optimistic_syncing/pipeline.go

```diff
@@ -12,77 +11,34 @@ import (
 	"github.com/onflow/flow-go/model/flow"
 )
 
-// State represents the state of the processing pipeline
-type State int
-
-const (
-	// StateReady is the initial state after instantiation and before downloading has begun
-	StateReady State = iota
-	// StateDownloading represents the state where data download is in progress
-	StateDownloading
-	// StateIndexing represents the state where data is being indexed
-	StateIndexing
-	// StateWaitingPersist represents the state where all data is indexed, but conditions to persist are not met
-	StateWaitingPersist
-	// StatePersisting represents the state where the indexed data is being persisted to storage
-	StatePersisting
-	// StateComplete represents the state where all data is persisted to storage
-	StateComplete
-	// StateCanceled represents the state where processing was aborted
-	StateCanceled
-)
-
-// String representation of states for logging
-func (s State) String() string {
-	switch s {
-	case StateReady:
-		return "ready"
-	case StateDownloading:
-		return "downloading"
-	case StateIndexing:
-		return "indexing"
-	case StateWaitingPersist:
-		return "waiting_persist"
-	case StatePersisting:
-		return "persisting"
-	case StateComplete:
-		return "complete"
-	case StateCanceled:
-		return "canceled"
-	default:
-		return ""
-	}
-}
-
-// StateUpdate contains state update information
-type StateUpdate struct {
-	// DescendsFromLastPersistedSealed indicates if this pipeline descends from
-	// the last persisted sealed result
-	DescendsFromLastPersistedSealed bool
-	// ParentState contains the state information from the parent pipeline
-	ParentState State
-}
-
 // StateUpdatePublisher is a function that publishes state updates
-type StateUpdatePublisher func(update StateUpdate)
+type StateUpdatePublisher func(resultID flow.Identifier, newState State)
 
 // Pipeline represents a generic processing pipeline with state transitions.
 // It processes data through sequential states: Ready -> Downloading -> Indexing ->
 // WaitingPersist -> Persisting -> Complete, with conditions for each transition.
-type Pipeline struct {
+type Pipeline interface {
+	Run(context.Context, Core) error
+	GetState() State
+	SetSealed()
+	OnParentStateUpdated(State)
```

Comment:
This is essentially carrying over information from one pipeline to another. I think that would be better placed in the `ResultsForest`. I guess all that is really relevant here is whether this pipeline is cancelled?

Reply:
also if the parent is in an active state which is required to start downloading